File: 프로그래머스/3/150367. 표현 가능한 이진트리/표현 가능한 이진트리.py
import sys
sys.setrecursionlimit(10**6)

def check(string):
    if len(string) == 1:
        return True
    center_idx = len(string) // 2
    if string[center_idx] == '0':
        if '1' in string:
            return False
    else:
        if len(string) == 3:
            return True
    if not check(string[:center_idx]) or not check(string[center_idx+1:]):
        return False
    
    return True


def solution(numbers):
    answer = []
    for num in numbers:
        bi_str = str(bin(num)[2:])
        length = len(bi_str)
        n = 0
        while True:
            if 2**n - 1 < length <= 2**(n+1) - 1:
                bi_str = ('0' * (2**(n+1) - 1 - length)) + bi_str
                break
            n += 1
        
        if check(bi_str):
            answer.append(1)
        else:
            answer.append(0)

    return answer

File: 프로그래머스/3/150367. 표현 가능한 이진트리/표현 가능한 이진트리.java
import java.lang.*;
import java.util.*;
class Solution {
    static boolean flag;
    public int[] solution(long[] numbers) {
        int[] answer = new int[numbers.length];
        int idx = 0;
        for(long num : numbers){
            String binary = Long.toBinaryString(num);
            int digit = 1;
            while(true){
                if(Math.pow(2, digit) - 1 >= binary.length()){
                    break;
                }
                digit++;
            }
            long pohawLen = (long)Math.pow(2, digit) - 1;
            while(pohawLen > binary.length()){
                binary = "0" + binary;
            }
            flag = true;
            check(binary);
            if(flag){
                answer[idx] = 1;
            }
            answer[idx] = flag ? 1 : 0;
            idx+=1;
        }
        return answer;
    }
    public static void check(String binary){
        if(!flag){
            return;
        }
        int mid = binary.length() / 2;
        if(binary.length() == 1){
            return;
        }
        if(binary.charAt(mid) == '0'){
            char[] list = binary.toCharArray();
            for(char c : list){
                if(c == '1'){
                    flag = false;
                    return;
                }
            }
        }
        if(binary.length() >= 3){
            check(binary.substring(0, mid));
            check(binary.substring(mid + 1));
        }
    }
}